internal/timex/timextest/fake_clock.go

```diff
@@ -0,0 +1,163 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Package timextest provides fake controlled time for testing code using the [timex] package. All methods here are safe
+// to call concurrently.
+package timextest
+
+import (
+	"sync"
+	"time"
+
+	"go.temporal.io/server/internal/timex"
+)
+
+type (
+	// FakeClock is a timex.Clock. Unlike other fake clock implementations, the methods of this clock are synchronous, so
+	// when you call Advance or Update, all triggered timers from AfterFunc will fire before the method returns, in the same
+	// goroutine.
+	FakeClock struct {
+		mu     sync.RWMutex
+		now    time.Time
+		timers []*fakeTimer
+	}
+
+	// fakeTimer is a fake implementation of [timex.Timer].
+	fakeTimer struct {
+		// need a link to the parent clock for synchronization
+		clock *FakeClock
+		// deadline for when the timer should fire
+		deadline time.Time
+		// callback to call when the timer fires
+		callback func()
+		// done is true if the timer has fired or been stopped
+		done bool
+		// index of the timer in the parent clock
+		index int
+	}
+)
+
+// NewFakeClock returns a FakeClock with the current time set to Unix zero: 1970-01-01 00:00:00 +0000 UTC.
+func NewFakeClock() *FakeClock {
+	return &FakeClock{
+		now: time.Unix(0, 0),
+	}
+}
+
+// Now return the current time.
+func (c *FakeClock) Now() time.Time {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	return c.now
+}
+
+// AfterFunc return a timer that will fire after the specified duration. It is important to note that the clock is
+// locked while the callback is called. This means that you must be cautious about calling any methods on the clock from
+// within the callback. Doing so will probably result in a deadlock. To avoid this, you may want to wrap all such calls
+// in a goroutine.
+func (c *FakeClock) AfterFunc(d time.Duration, f func()) timex.Timer {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	t := &fakeTimer{clock: c, deadline: c.now.Add(d), callback: f}
+	t.index = len(c.timers)
+	c.timers = append(c.timers, t)
+
+	return t
+}
+
+// Update the fake current time. It returns the clock so that you can chain calls like this:
+// clock := NewFakeClock().Update(time.Now())
+func (c *FakeClock) Update(now time.Time) *FakeClock {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.now = now
+	c.fireTimers()
+	return c
+}
+
+// Advance the timer by the specified duration.
+func (c *FakeClock) Advance(d time.Duration) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.now = c.now.Add(d)
+	c.fireTimers()
+}
+
+// update the fake current time without locking.
+
+// fireTimers fires all timers that are ready.
+func (c *FakeClock) fireTimers() {
+	n := 0
+	for _, t := range c.timers {
```

Comment:
Now that I think about it, you should probably fire timers in order of first deadline, but it probably doesn't matter